File: 1/2020.09.29/104.Maximun Depth of Binary Tree/java/solution.java
class Solution {
    public int maxDepth(TreeNode root) {
        // DFS, recursively
        if (root == null) {
            return 0;
        }
        return Math.max(maxDepth(root.left), maxDepth(root.right)) + 1;
    }

    public int maxDepth(TreeNode root) {
        // BFS
        int depth = 0;
        if (root == null) {
            return 0;
        }
        Deque<TreeNode> q = new LinkedList<>();
        q.offerLast(root);
        while (!q.isEmpty()) {
            int size = q.size();
            while (size-- > 0) {
                TreeNode node = q.pollFirst();
                if (node.left != null) {
                    q.offerLast(node.left);
                }
                if (node.right != null) {
                    q.offerLast(node.right);
                }
            }
            depth++;
        }
        return depth;
    }
}

File: 1/2020.09.29/104.Maximun Depth of Binary Tree/python/solution.py
class Solution:
    def maxDepth(self, root: TreeNode) -> int:
        # DFS, recursively
        if not root:
            return 0
        return max(self.maxDepth(root.left), self.maxDepth(root.right)) + 1

    def maxDepth(self, root: TreeNode) -> int:
        # BFS
        depth = 0
        if not root:
            return depth
        q = [root]
        while q != []:
            n = len(q)
            for _ in range(n):
                node = q.pop(0)
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
            depth += 1
        return depth
